File: front-end/src/app/domain/models/game/action/impl/money/MoneyExchangeAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";

/**
 * An abstract representation, of the action, of money exchange among two parties within a GameSession,
 * with at least one being an instance of Player. The (fictional monopoly) bank being represented by null.
 *
 * @author Hamza el Haouti
 */
export abstract class MoneyExchangeAction extends GameAction {

  protected constructor(
    classType: string,
    id: number,
    payee: Player,
    performedAt: Date,
    recipient: Player,
    amount: number
  ) {
    super(classType, id, payee, performedAt);
    this._recipient = recipient;
    this._amount = amount;
  }

  private _recipient: Player;

  public get recipient(): Player {
    return this._recipient;
  }

  private _amount: number;

  public get amount(): number {
    return this._amount;
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "recipient": this.recipient?.toJson(),
        "amount": this.amount,
      }
    );
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/MoneyExchangeAction.java
package nl.hva.backend.domain.models.game.action.impl.money;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

/**
 * An abstract representation, of the action, of money exchange among two parties within a GameSession,
 * with at least one being an instance of Player. The (fictional monopoly) bank being represented by null.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public abstract class MoneyExchangeAction extends GameAction {

    @ManyToOne
    private Player recipient;

    private double amount;

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param payee     The party paying, a Player instance, or null.
     * @param recipient The receiving party, a Player instance, or null.
     * @param amount    The amount, to be subtracted from the payee, and added to the balance of the recipient.
     */
    protected MoneyExchangeAction(
            Player payee,
            Player recipient,
            double amount
    ) {
        super(payee);
        this.recipient = recipient;
        this.amount = amount;
    }

}